(closes #2739) Update PSyKAl transformation script signature to accept PSYIR node

sergisiso opened the pull request:
@hiker suggestion (https://github.com/stfc/PSyclone/issues/2739#issuecomment-2402167266) of having a temporary backwards compatible property for the psy.invokes... worked very well. And I agree having a grace period when psyclone2-style signatures are accepted is a good strategy. So LFRic and PSycloneBench can be updated without rushing it.

The only change needed was that PSyKAl APIs only had 1 container on top of the routines, but for the code that it was representing the proper PSyIR is FileContainer->Container->InvokeSchedules. So I had to add it and update a few symbol_table links to make it all work.

With the minimal changes in the first commit of this PR, both signature styles are supported.

Nevertheless, I will update all the examples in the documentation and tutorials to the new style also in this PR to make clear what is the intended format going forward.

sergisiso commented:
@arporter @hiker This is ready for review. As mentioned above, even though this still supports the old format, I updated all the scripts in the repository. I also launched the integration tests.

In addition to the input argument I also removed some `print(psyir.view)` and all `return psy | psyir` from the scripts. The reason for the first is that I think they are now too long and its easier to look at the resulting Fortran to see if they did what is supposed. Regarding the "return" I actually like the ability to chain functions using the dot operator, however the lowering and generation do not always take the returned value of the script and therefore if the return is an object different from the received one, it doesn't do what one would expect. So I deleted it, for now, to avoid being misleading.

arporter commented on tutorial/practicals/LFRic/single_node/2_openacc/acc_parallel.py:
```diff
@@ -45,28 +45,21 @@
 
-def trans(psy):
+def trans(psyir):
     '''PSyclone transformation script for the dynamo0p3 api to apply
```
"LFRic"

sergisiso replied:
Done

arporter reviewed (changes requested):
Thanks very much Sergi, I think this is a really good improvement. Almost all of my comments are just about doc strings. There are some clashes to resolve too.
I'll check the integration tests next time.

sergisiso commented on doc/user_guide/getting_going.rst:
```diff
@@ -566,6 +566,26 @@ with a `trans` function defined. For example:
+           for invoke in psy.invokes.invoke_list:
+              invoke.schedule
+
+   This is deprecated and will stop working in the upcoming psyclone release.
```
Done

sergisiso commented on doc/user_guide/getting_going.rst:
```diff
@@ -566,6 +566,26 @@ with a `trans` function defined. For example:
+       def trans(psy):
+           ''' Add OpenMP Parallel Loop directives.
+
+           :param psy: the PSy object that PSyclone has constructed for the \
```
Removed

sergisiso commented on doc/user_guide/psy_data.rst:
```diff
@@ -117,25 +117,21 @@ be printed at runtime, e.g.::
+:ref:`GOcean API <gocean-api>` - no API-specific transformations are required.
+Below is an example that searches for each loop in a PSyKAl invoke code (which
+will always surround kernel calls) and applies the transformation to each one.
+This code has been successfully used as aglobal transformation with the LFRic
```
Fixed

sergisiso commented on doc/user_guide/psy_data.rst:
```diff
@@ -117,25 +117,21 @@ be printed at runtime, e.g.::
+Below is an example that searches for each loop in a PSyKAl invoke code (which
+will always surround kernel calls) and applies the transformation to each one.
+This code has been successfully used as aglobal transformation with the LFRic
+Gravity Wave miniapp (the executable is named ``gravity_wave``)
```
Done

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -774,26 +774,25 @@ what the **psyclone** script calls internally).
 
-A valid script file must contain a **trans** function which accepts a **PSy**
-object as an argument and returns a **PSy** object, i.e.:
+A valid script file must contain a **trans** function which accepts a **PSyIR**
```
Done

sergisiso replied:
(Oh and I noticed that the code just above was not rendered properly - I assumed the triple has was meant to comment it out but didn't work, so I delete it)

sergisiso commented on examples/gocean/eg1/opencl_transformation.py:
```diff
@@ -36,22 +36,17 @@
     Transformation routine for use with PSyclone. Converts any imported-
     variable accesses into kernel arguments and then applies the OpenCL
     transformation to the PSy layer.
 
```
Done

sergisiso commented on examples/gocean/eg1/openmp_taskloop_trans.py:
```diff
@@ -39,37 +39,27 @@
+def trans(psyir):
     '''
     Transformation routine for use with PSyclone. Applies the OpenMP
     taskloop and taskwait transformations to the PSy layer.
```
Done

sergisiso commented on examples/gocean/eg2/acc_transform.py:
```diff
@@ -43,37 +43,32 @@
+                ktrans.apply(kern)
+                # Ideally we would module-inline the kernel here (to save
+                # having to rely on the compiler to do it) but this does not
+                # currently work for the fparser2 AST (issue #229).
```
Done

sergisiso commented on examples/gocean/eg4/acc_transform.py:
```diff
@@ -45,19 +45,17 @@
-    to the schedule of the first invoke and return the new psy object '''
+def trans(psyir):
+    ''' Take the supplied psy-layer, apply OpenACC transformations
+    to the schedule of the first invoke.'''
```
Done

sergisiso commented on examples/gocean/eg5/extract/extract_transform.py:
```diff
@@ -44,33 +44,23 @@
-def trans(psy):
+def trans(psyir):
     '''
     Take the supplied psy object, and add kernel extraction code.
```
Done

sergisiso commented on examples/gocean/eg5/readonly/read_only_transform.py:
```diff
@@ -40,37 +40,26 @@
-    :returns: the transformed PSy object.
-    :rtype: :py:class:`psyclone.gocean1p0.GOPSy`
+    Add verification checks to both invokes that read only parameters are not
+    modified.
```
Done

sergisiso commented on examples/gocean/eg6/backends_transform.py:
```diff
@@ -44,11 +44,11 @@
 
-def trans(psy):
+def trans(psyir):
     ''' Use the PSyIR back-end to generate PSy-layer target code'''
```
Yeah, I am tempted to remove it but it would leave a hole and would need to rename many folders, so I will not do it in this PR. But there are a few examples with old stuff that could be clean up.

For now I did what you suggested.

sergisiso commented on examples/lfric/eg11/async_script.py:
```diff
@@ -41,7 +41,7 @@
+def trans(psyir):
     '''A sample transformation script to demonstrate the use of asynchronous
     halo exchanges with overlapping compute and communication for the
     most costly halo exchanges in the (current version of the) LFRic model.
```
Done

sergisiso commented on examples/lfric/eg12/colouring_and_omp.py:
```diff
@@ -39,42 +39,35 @@
-def trans(psy):
+def trans(psyir):
     ''' PSyclone transformation script for the dynamo0p3 api to apply
     colouring and OpenMP generically.'''
```
Done

sergisiso commented on examples/lfric/eg12/extract_kernel_with_transformations.py:
```diff
@@ -77,7 +78,7 @@
 
-def trans(psy):
+def trans(psyir):
     ''' PSyclone transformation script for the Dynamo0.3 API to
```
Done

sergisiso commented on examples/lfric/eg12/extract_nodes.py:
```diff
@@ -71,23 +70,18 @@
-def trans(psy):
+def trans(psyir):
     ''' PSyclone transformation script for the Dynamo0.3 API to extract
     the specified Nodes in an Invoke. '''
```
Done

sergisiso commented on examples/lfric/eg13/kernel_constants.py:
```diff
@@ -72,24 +71,19 @@
 
-def trans(psy):
+def trans(psyir):
     '''PSyclone transformation script for the Dynamo0.3 API to make the
```
Done

sergisiso commented on examples/lfric/eg15/matvec_opt.py:
```diff
@@ -74,7 +74,7 @@
 
-def trans(psy):
+def trans(psyir):
     '''PSyclone transformation script for the Dynamo0.3 API to optimise
```
Done

sergisiso commented on examples/lfric/eg2/print_psyir_trans.py:
```diff
@@ -34,30 +34,18 @@
 
 
-def trans(psy):
+def trans(psyir):
```
Done

sergisiso commented on examples/lfric/eg2/print_psyir_trans.py:
```diff
@@ -34,30 +34,18 @@
 # Author R. W. Ford, STFC Daresbury Laboratory
 
 ''' Simple transformation script that prints out the names of the 'invoke'(s)
```
I made some changes in the docstrings

sergisiso commented on examples/lfric/eg4/backends_transform.py:
```diff
@@ -40,31 +40,27 @@
 
-def trans(psy):
+def trans(psyir):
     ''' Use the PSyIR back-end to generate PSy-layer target code'''
```
Done

sergisiso commented on examples/lfric/eg4/backends_transform.py:
```diff
@@ -40,31 +40,27 @@
-    print("f2pygen code:")
-    print(str(psy.gen))
-
     # TODO #1010: This script should terminate here until LFRic declares
```
Yes, but this example probably should be removed altogether once that is done, as this is what will happen after the script is executed. I think I created this a long time ago to see track the progress.

sergisiso commented on examples/lfric/eg6/omp_reprod_script.py:
```diff
@@ -45,15 +45,15 @@
-def trans(psy):
+def trans(psyir):
     ''' PSyclone transformation script for the dynamo0p3 API to apply
     loop fusion and OpenMP for a particular example.'''
```
Done

sergisiso commented on examples/lfric/eg6/omp_script.py:
```diff
@@ -45,14 +45,14 @@
-def trans(psy):
+def trans(psyir):
     ''' PSyclone transformation script for the dynamo0p3 API to apply
     loop fusion and OpenMP for a particular example.'''
```
Done

sergisiso commented on examples/lfric/eg8/redundant_script.py:
```diff
@@ -46,22 +46,21 @@
-def trans(psy):
+def trans(psyir):
     '''removes the grad_p halo exchanges by redundant computation then
     moves the remaining halo exchanges to the beginning of the invoke
```
Done

sergisiso commented on examples/lfric/eg9/colouring_and_omp.py:
```diff
@@ -39,42 +39,36 @@
 ''' File containing a PSyclone transformation script for the Dynamo0p3
```
Done

sergisiso commented on examples/lfric/scripts/async_halo_exchanges.py:
```diff
@@ -42,37 +42,34 @@
-def trans(psy):
+def trans(psyir):
     '''A transformation script to use asynchronous halo exchanges with
     overlapping compute and communication for the LFRic model. '''
```
Done

sergisiso commented on examples/lfric/scripts/colouring_and_omp.py:
```diff
@@ -39,42 +39,35 @@
 ''' File containing a PSyclone transformation script for the Dynamo0p3
```
Done

sergisiso commented on examples/lfric/scripts/colouring_and_omp.py:
```diff
@@ -39,42 +39,35 @@
 
-def trans(psy):
+def trans(psyir):
     ''' PSyclone transformation script for the dynamo0p3 api to apply
```
Done

sergisiso commented on examples/lfric/scripts/everything_everywhere_all_at_once.py:
```diff
@@ -61,7 +62,7 @@
 
-def trans(psy):
+def trans(psyir):
     ''' Apply all possible LFRic transformations. '''
```
Done

sergisiso commented on examples/lfric/scripts/everything_everywhere_all_at_once.py:
```diff
@@ -73,72 +74,65 @@ def trans(psy):
-        for kschedule in root.walk(KernelSchedule):
+            # Transformations that modify kernel code will need to have the
+            # kernels inlined first
             if ENABLE_INTRINSIC_INLINING:
```
The idea of `ENABLE_INTRINSIC_INLINING` is to enable the Matmul2Code below and the name is still valid.

This line here is just a pre-step to this, but the idea was that other optimisations may also benefit from inlining the kernels first, so this condition that you mark would grow with the other enabled features, but not necessarily mentioning the kernel inlining itself.

arporter replied:
OK, I noticed that actually, it did make more sense lower down.

sergisiso commented on examples/lfric/scripts/gpu_offloading.py:
```diff
@@ -61,7 +61,7 @@
+def trans(psyir):
     '''Applies PSyclone colouring and GPU offloading transformations. Any
     kernels that cannot be offloaded to GPU are parallelised using OpenMP
     on the CPU. Any setval_* kernels are transformed so as to compute
```
Done

sergisiso commented on examples/lfric/scripts/inline_kernels_and_intrinsics.py:
```diff
@@ -39,44 +39,36 @@
 
-def trans(psy):
+def trans(psyir):
     '''Applies LFRic serial optimisations'''
```
Done

sergisiso commented on examples/lfric/scripts/inline_kernels_and_intrinsics.py:
```diff
@@ -39,44 +39,36 @@
+    for kernel in psyir.coded_kernels():
+        try:
+            inline_trans.apply(kernel)
+            print(f"Inline transformation was successful for "
```
Agreed, done

sergisiso commented on examples/lfric/scripts/loop_fuse.py:
```diff
@@ -43,40 +43,29 @@
+def trans(psyir):
+    '''PSyclone transformation script for the LFRic API to apply loop
     fusion generically to all top level loops.
 
```
Done

sergisiso commented on examples/lfric/scripts/redundant_dofs.py:
```diff
@@ -31,21 +31,20 @@
-# Authors: R. W. Ford and N. Nobre, STFC Daresbury Lab
+# Authors: R. W. Ford, N. Nobre and S. Siso, STFC Daresbury Lab
 
 '''File containing a PSyclone transformation script for the Dynamo0.3
```
Done

sergisiso commented on examples/lfric/scripts/redundant_dofs.py:
```diff
@@ -56,20 +55,17 @@ def trans(psy):
+        if loop.iteration_space in ITERATION_SPACES:
+            # we may have more than one kernel in the loop so
+            # check that none of them are reductions
+            reduction = False
```
Good idea

sergisiso commented on examples/lfric/scripts/redundant_setval_c.py:
```diff
@@ -31,22 +31,21 @@
+def trans(psyir):
     '''PSyclone transformation script for the lfric API to apply
     redundant computation into the level 1 halo generically to all
     loops that iterate over dofs and exclusively contain the setval_c
```
Done

sergisiso commented on examples/lfric/scripts/redundant_setval_c.py:
```diff
@@ -60,21 +59,18 @@ def trans(psy):
+            # we may have more than one kernel in the loop so
+            # check that all of them are in the list of accepted
+            # kernel names
+            setcalls = True
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -188,12 +188,12 @@ def gen_code(self, parent):
             # Then insert it into a f2pygen AST as a PSyIRGen node.
             # Note that other routines besides the Invoke could have been
             # inserted during the lowering (e.g. module-inlined kernels),
             # so have to iterate over all current children of root.
```
Fixed

sergisiso commented on src/psyclone/psyir/nodes/file_container.py:
```diff
@@ -55,6 +57,46 @@ class FileContainer(Container):
+    def invokes(self):
+        ''' Return the Invokes object associated to this FileContainer.
+        This is for compatibility with old psyclone transformation scripts.
+        Before the entry point was PSy, and the script had to find the list
```
Done

sergisiso commented on src/psyclone/psyir/nodes/file_container.py:
```diff
@@ -55,6 +57,46 @@ class FileContainer(Container):
+        '''
+        print("Deprecation warning: PSyclone script uses the legacy "
+              "transformation signature 'def trans(psy)', please update the "
+              "script to recive the root psyir node as argument.",
```
Fixed

arporter replied:
Not quite :-) I'll do this if it's the only thing left.

sergisiso replied:
Ups, done now

sergisiso commented on src/psyclone/psyir/nodes/file_container.py:
```diff
@@ -55,6 +57,46 @@ class FileContainer(Container):
+            def trans(psy):
+                psy.invokes.get_invoke('name').schedule
+
+        Still work as expected. However, it exposes the PSy hierachy to
```
Done

sergisiso commented on src/psyclone/tests/psyir/nodes/file_container_test.py:
```diff
@@ -97,3 +98,32 @@ def test_file_container_create():
+    # Get a psykal FileContainer
+    filecontainer = invoke.schedule.root
+
+    # This can be used as if it where a PSy
```
Fixed

sergisiso commented on src/psyclone/tests/psyir/nodes/file_container_test.py:
```diff
@@ -97,3 +98,32 @@ def test_file_container_create():
+    filecontainer = FileContainer("test")
+    with pytest.raises(NoInvokesError) as err:
+        _ = filecontainer.invokes
+    assert ("No InvokeSchedule found in 'test', does it come from a "
```
Done

sergisiso commented on src/psyclone/psyir/nodes/file_container.py:
```diff
@@ -55,6 +57,46 @@ class FileContainer(Container):
+        Before the entry point was PSy, and the script had to find the list
+        of InvokeSchedules, now the entry point is the root FileContainer:
+
+        before: PSy -> Invokes -> Invoke -> InvokeSchedule
```
Done (apparently it always need indentation after a sentence ending in colon)

sergisiso commented on tutorial/practicals/LFRic/building_code/4_psydata/README.md:
```diff
@@ -40,14 +40,13 @@ is to use the name that was given to the invoke call. Look at the
 file ``time_evolution_alg_mod.x90`` and find the name given to the
 invoke statement that propagates the perturbation.
 
```
I am now pointing to the TODOs in the script, instead of specific lines, is that ok?

arporter replied:
Great, that's a good idea :-)

sergisiso commented on tutorial/practicals/LFRic/building_code/4_psydata/solutions/extract_all_transform.py:
```diff
@@ -33,44 +33,34 @@
-def trans(psy):
+def trans(psyir):
     '''
     Take the supplied psy object, and add kernel extraction code.
```
Done

sergisiso commented on tutorial/practicals/LFRic/distributed_memory/4_reductions/schedule.py:
```diff
@@ -32,6 +32,7 @@
+# Modified: S. Siso, STFC Daresbury Lab
 
 '''A PSyclone transformation script that outputs a textual
 representation of the PSyIR representing the PSy-layer for the first
```
Done

sergisiso commented on tutorial/practicals/LFRic/single_node/1_openmp/omp_script.py:
```diff
@@ -31,45 +31,38 @@
 
-def trans(psy):
+def trans(psyir):
     '''PSyclone transformation script for the dynamo0p3 API that applies
```
Done

sergisiso commented on src/psyclone/tests/test_files/gocean1p0/legacy_script.py:
```diff
@@ -0,0 +1,52 @@
+# -----------------------------------------------------------------------------
```
Done, I used a factory pattern

arporter replied:
Nice. It's good to be able to see the content of the script in the code performing the test.

sergisiso commented:
@arporter This is ready for another review

arporter commented on src/psyclone/tests/generator_test.py:
```diff
@@ -80,13 +80,32 @@
+    def populate_script(string):
+        with open(tmpfile, 'w+', encoding="utf8") as script:
+            script.write(string)
+        return tmpfile
```
This returns `tmpfile` which is created outside of `populate_script()`. Is that OK?

sergisiso replied:
Yes, it is a [Closure](https://www.programiz.com/python-programming/closure). In the returned function I don't want it as a parameter but embedded inside fixed to the value provided by the fixture.

arporter replied:
Aha. I live and learn :-)

arporter commented on src/psyclone/tests/generator_test.py:
```diff
@@ -142,66 +161,77 @@ def test_script_file_wrong_extension():
     '''
-    with pytest.raises(Exception) as error_syntax:
+    error_syntax = script_factory("""
+this is invalid python
```
Much nicer now (IMHO) :-)

arporter commented on src/psyclone/tests/generator_test.py:
```diff
@@ -210,36 +240,42 @@ def test_script_no_trans_alg(capsys):
     assert ("Deprecation warning: PSyclone script uses the legacy "
             "transformation signature 'def trans(psy)', please update the "
-            "script to recive the root psyir node as argument."
+            "script to recieve the root psyir node as argument."
```
"receive" ('i before e except after c'!)

sergisiso replied:
Fixed

arporter commented on src/psyclone/tests/generator_test.py:
```diff
@@ -417,41 +460,43 @@ def test_profile_gocean():
-    in a valid script file does no transformations (it simply passes
-    input to output). In this case the valid script file has an
-    explicit path and must therefore exist at this location.
+    in a valid script file does no transformations. In this case the valid
```
Can probably remove the last sentence as there's no path manipulation here any more.

arporter reviewed (changes requested):
Thanks Sergi, pretty much there now. Just a bit of tidying required and I have a question about the creation of the file object in the new fixture that you've added.

Also, not strictly this PR (I think) but the Ref. Guide is up to 19 warnings now:
```
PSyclone/src/psyclone/domain/lfric/lfric_kern.py:docstring of psyclone.domain.lfric.LFRicKern.gen_stub:6: WARNING: Inline emphasis start-string without end-string.
```
Would you mind taking a look (unless you've already fixed this in your documentation-related PR)?

arporter commented:
(I can see that the integration tests worked last time once you'd fixed the compilation tests..)

sergisiso commented:
@arporter I addressed the comments and brought it to master. This is ready for another look.

The only thing not fixed is:

> Would you mind taking a look (unless you've already fixed this in your documentation-related PR)?

That PR is for all but the reference guide unfortunately which are more difficult to resolve. However, I am not getting the warning you mentioned with Sphinx 7.2.6. Maybe its a version thing as I get 21 warnings instead of 19 but all seem related to #1280 ?

arporter reviewed (approved):
All requested changes have been made, thanks.
FWIW, I'm using Sphinx 7.4.7. This is the most recent version that fulfils<8 as required by sphinx-rtd-theme.
Will proceed to merge.

arporter commented:
I've just realised I built the Ref. Guide on a different branch! Ahem. If I do it on this branch then I'm back to 17 warnings :-)
